packages/ddp-server/writefence.js

```diff
@@ -59,6 +60,15 @@ _.extend(DDPServer._WriteFence.prototype, {
     self._maybeFire();
   },
 
+  // Register a function to be called before firing the fence.
+  onBeforeFire: function (func) {
+    var self = this;
+    if (self.fired)
```

Comment:
It will **not** be correct to check for `self.armed` instead. Meteor.refresh might get called for a fence after it has been armed in case if method issued async writes (by passing a callback to database methods). That was the reason I have moved it to 'before fire' instead of 'on arm'.

I have added additional code to ensure that everything is re-entrant - if for any reason another write is scheduled on the fence when onBeforeFire callback is being processed, it will attach a new callback and catch up to oplog again. The method has to do really funky stuff to get to that situation though - like doing writes from setTimeout callback and doing 'normal' writes as well.